projects/gateway2/deployer/values_helpers.go

```diff
@@ -22,43 +23,64 @@ var ComponentLogLevelEmptyError = func(key string, value string) error {
 	return eris.Errorf("an empty key or value was provided in componentLogLevels: key=%s, value=%s", key, value)
 }
 
-// Extract the listener ports from a Gateway. These will be used to populate:
+// Extract the listener ports from a Gateway and corresponding listener sets. These will be used to populate:
 // 1. the ports exposed on the envoy container
 // 2. the ports exposed on the proxy service
-func getPortsValues(gw *api.Gateway, gwp *v1alpha1.GatewayParameters) []helmPort {
+func getPortsValues(cgw *types.ConsolidatedGateway, gwp *v1alpha1.GatewayParameters) []helmPort {
 	gwPorts := []helmPort{}
-	for _, l := range gw.Spec.Listeners {
-		listenerPort := uint16(l.Port)
-
-		// only process this port if we haven't already processed a listener with the same port
-		if slices.IndexFunc(gwPorts, func(p helmPort) bool { return *p.Port == listenerPort }) != -1 {
-			continue
+	for _, l := range cgw.Gateway.Spec.Listeners {
+		gwPorts = appendPortValue(gwPorts, uint16(l.Port), string(l.Name), gwp)
+	}
+	for i, ls := range cgw.AllowedListenerSets {
+		for _, l := range ls.Spec.Listeners {
+			gwPorts = appendPortValue(gwPorts, uint16(l.Port), fmt.Sprintf("%d-%s", i, l.Name), gwp)
 		}
+	}
+	return gwPorts
+}
 
-		targetPort := ports.TranslatePort(listenerPort)
-		portName := string(l.Name)
-		protocol := "TCP"
-
-		// Search for static NodePort set from the GatewayParameters spec
-		// If not found the default value of `nil` will not render anything.
-		var nodePort *uint16 = nil
-		if gwp.Spec.GetKube().GetService().GetType() != nil && *(gwp.Spec.GetKube().GetService().GetType()) == corev1.ServiceTypeNodePort {
-			if idx := slices.IndexFunc(gwp.Spec.GetKube().GetService().GetPorts(), func(p *v1alpha1.Port) bool {
-				return p.GetPort() == uint16(listenerPort)
-			}); idx != -1 {
-				nodePort = ptr.To(uint16(*gwp.Spec.GetKube().GetService().GetPorts()[idx].GetNodePort()))
-			}
-		}
+func sanitizePortName(name string) string {
+	nonAlphanumericRegex := regexp.MustCompile(`[^a-zA-Z0-9-]+`)
+	str := nonAlphanumericRegex.ReplaceAllString(name, "-")
+	doubleHyphen := regexp.MustCompile(`-{2,}`)
+	str = doubleHyphen.ReplaceAllString(str, "-")
 
-		gwPorts = append(gwPorts, helmPort{
-			Port:       &listenerPort,
-			TargetPort: &targetPort,
-			Name:       &portName,
-			Protocol:   &protocol,
-			NodePort:   nodePort,
-		})
+	// This is a kubernetes spec requirement.
+	maxPortNameLength := 15
+	if len(str) > maxPortNameLength {
+		str = str[:maxPortNameLength]
 	}
-	return gwPorts
+	return str
+}
+
+func appendPortValue(gwPorts []helmPort, port uint16, name string, gwp *v1alpha1.GatewayParameters) []helmPort {
+	// only process this port if we haven't already processed a listener with the same port
+	if slices.IndexFunc(gwPorts, func(p helmPort) bool { return *p.Port == port }) != -1 {
```

Comment:
nit: this is n^2 when called from `getPortsValues`, could use a set.

Reply:
Updated to be more optimized

Reply:
Can deal with it a follow up. Maybe we do this earlier in the building of ConsolidatedGateway and keep the overlapped/ignored ones in Rejected?